spec/controllers/work_packages/types/subject_configuration_controller_spec.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+# -- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+#
+
+require "spec_helper"
+
+RSpec.describe WorkPackages::Types::SubjectConfigurationController do
+  let(:user) { create(:admin) }
+  let(:wp_type) { create(:type) }
+
+  current_user { user }
+
+  context "when the user is not logged in" do
+    let(:user) { User.anonymous }
+
+    it "responds with forbidden" do
+      put :update_subject_configuration, params: { id: wp_type.id }
+      expect(response.status).to redirect_to signin_url(back_url: subject_configuration_type_url(wp_type))
+    end
+  end
+
+  context "when the user is not an admin" do
+    let(:user) { create(:user) }
+
+    it "responds with forbidden" do
+      put :update_subject_configuration, params: { id: wp_type.id }
+      expect(response).to have_http_status :forbidden
+    end
+  end
+
+  describe "PUT #update_subject_configuration" do
```

Comment:
:yellow_circle: Both contexts (when the user is not an admin and when the user is not logged in) above are testing the behaviour of the `update_subject_configuration` action when receiving a PUT. I think those contexts should semantically be placed inside this `describe` as well.

As far as I can tell, this should also not conflict with the additional setup peformed in the current describe block.

Reply:
While this is true, that it probably won't clash, the two contexts above are about controller before_actions. So, technically, those are not about this PUT, we just need any action to trigger the before action. Would you be fine to leave them outside?

Reply:
Let me add this perspective: That these things are `before_action`s is an implementation detail. A `before_filter` can take an `only` or `except` argument, there is no guarantee that it runs for each action.

So I'd probably ask a proper unit test for a controller to repeat this expectation for all of the controller actions, to ensure that all actions behave the way they should. That's also the reason why I saw this context inside the action `describe`.

I am definitely not going to die on that hill, i.e. I'd be fine to leave the context where it is (especially considering that we only have one action so far).